ext/drx/drx.c

```diff
@@ -1372,3 +1405,1117 @@ drx_tail_pad_block(void *drcontext, instrlist_t *ilist)
+                                                   sg_info->scale, sg_info->disp, OPSZ_8),
+                             opnd_create_reg(scalar_value_reg)),
+                         orig_app_pc));
+    } else {
+        ASSERT(false, "Unexpected index size.");
+        return false;
+    }
+    return true;
+}
+
+static bool
+expand_gather_load_scalar_value(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
+                                scatter_gather_info_t *sg_info, reg_id_t scalar_index_reg,
+                                app_pc orig_app_pc)
+{
+    if (sg_info->base_reg == IF_X64_ELSE(DR_REG_RAX, DR_REG_EAX)) {
+        /* We need the app's base register value. If it's xax, then it may be used to
+         * store flags by drreg.
+         */
+        drreg_get_app_value(drcontext, bb, sg_instr, sg_info->base_reg,
+                            sg_info->base_reg);
+    }
+    if (sg_info->scalar_value_size == OPSZ_4) {
+        PREXL8(
+            bb, sg_instr,
+            INSTR_XL8(INSTR_CREATE_mov_ld(
+                          drcontext,
+                          opnd_create_reg(IF_X64_ELSE(reg_64_to_32(scalar_index_reg),
+                                                      scalar_index_reg)),
+                          opnd_create_base_disp(sg_info->base_reg, scalar_index_reg,
+                                                sg_info->scale, sg_info->disp, OPSZ_4)),
+                      orig_app_pc));
+    } else if (sg_info->scalar_value_size == OPSZ_8) {
+        ASSERT(reg_is_64bit(scalar_index_reg),
+               "Internal error: scratch register not 64-bit.");
+        PREXL8(bb, sg_instr,
+               INSTR_XL8(INSTR_CREATE_mov_ld(drcontext, opnd_create_reg(scalar_index_reg),
+                                             opnd_create_base_disp(
+                                                 sg_info->base_reg, scalar_index_reg,
+                                                 sg_info->scale, sg_info->disp, OPSZ_8)),
+                         orig_app_pc));
+    } else {
+        ASSERT(false, "Unexpected index size.");
+        return false;
+    }
+    return true;
+}
+
+#endif
+
+/* The function expand scatter and gather instructions to a sequence of equivalent scalar
+ * operations. Gather instructions are expanded into a sequence of mask register bit
+ * tests, extracting the index value, a scalar load, inserting the scalar value into the
+ * destination simd register, and mask register bit updates. Scatter instructions are
+ * similarly expanded into a sequence, but using a scalar store.
+ *
+ * AVX2 vpgatherdd, vgatherdps, vpgatherdq, vgatherdpd, vpgatherqd, vgatherqps,
+ * vpgatherqq, vgatherqpd:
+ *
+ * vpgatherdd (%rax,%ymm1,4)[4byte] %ymm2 -> %ymm0 %ymm2 sequence laid out here, others
+ * are similar:
+ *
+ * Extract mask dword. qword versions use vpextrq:
+ *
+ * vextracti128 %ymm2 $0x00 -> %xmm3
+ * vpextrd %xmm3 $0x00 -> %ecx
+ *
+ * Test mask bit:
+ *
```

Comment:
I would also get rid of these blank lines and tighten this up so it's not so long: hard to read it when it fills so many screens.  With indenting the blank lines are not needed and it will be shorter and easier to look at.

Reply:
Formatting, "pretty-izing" it now

Reply:
Done, please check.